### ISSUE TITLE
Problem with automatic RNG scoping

### ISSUE THREAD

**Issue by romainfrancois** [RcppCore/Rcpp#268]
Consider: 

``` cpp
#include <Rcpp.h>
using namespace Rcpp ;

// [[Rcpp::export]]
List test_rnorm(){
    Function r_rnorm("rnorm") ;

    NumericVector x = rnorm(2) ;
    NumericVector y = r_rnorm(2) ;

    return List::create(x, y );

}

/*** R 
    test_rnorm()
*/
```

I get: 

```
> test_rnorm()
[[1]]
[1] -0.1906103  0.5690450

[[2]]
[1] -0.1906103  0.5690450
```

Obviously because the seed is not `put` between the `C++` call to `Rcpp::rnorm` and the `rnorm` function from R. 

Of course this is trivial in this example because I'm just calling `rnorm`, but this could have been calling some other R function that internally does the `get` and `put` dance. 

Going further, if we consider this one now: 

``` cpp
#include <Rcpp.h>
using namespace Rcpp ;

// [[Rcpp::export]]
List test_rnorm( int ncpp, int nr ){
    Function r_rnorm("rnorm") ;

    NumericVector x = rnorm(ncpp) ;
    NumericVector y = r_rnorm(nr) ;

    return List::create(x, y );

}

/*** R 
    test_rnorm(4, 2); rnorm(2)
*/
```

which gives: 

```
> test_rnorm(4, 2)
[[1]]
[1]  0.9418597  2.0877568  0.9308808 -0.2713577

[[2]]
[1] 0.9418597 2.0877568


> rnorm(2)
[1]  0.9308808 -0.2713577
```

So I think implicit automatic rng scoping is wrong. Not sure about a workaround though. I guess reverting to explicit uses of `RNGScope` can't be done, etc ...

**Comment by eddelbuettel** [comment]
Nice example.

But my first reaction is what I would say to someone who claims `unlink()` is dangerous...

To me, this seems mostly like a documentation issue as it is indeed hard to safeguard against bad use.  We give three different ways to get to the RNGs from the C++ level; if folks then insist on bypassing the carefully constructed framework they will have to live with the consequences of their actions.    That said, if we had a way to warn more directly / make it more difficult I would be in favour too.

**Comment by eddelbuettel** [comment]
Also note that @jjallaire added `rng=false` as an option.  If we do that, things still work -- but of course not in the corner case of forcing `RNGscope` off, yet still mixing with R calls as in your example:

```{.r}
R> sourceCpp("/tmp/romain.cpp", verbose=TRUE, rebuild=TRUE)

 Generated extern "C" functions
 --------------------------------------------------------


 #include <Rcpp.h>                                                                                                                                                                                                                        
 // test_rnorm
 List test_rnorm(int ncpp, int nr);
 RcppExport SEXP sourceCpp_13765_test_rnorm(SEXP ncppSEXP, SEXP nrSEXP) {
 BEGIN_RCPP
     Rcpp::RObject __result;
     Rcpp::traits::input_parameter< int >::type ncpp(ncppSEXP);
     Rcpp::traits::input_parameter< int >::type nr(nrSEXP);
     __result = Rcpp::wrap(test_rnorm(ncpp, nr));
     return __result;
 END_RCPP
 }
 // test_rnorm2
 NumericVector test_rnorm2(int nr);
 RcppExport SEXP sourceCpp_13765_test_rnorm2(SEXP nrSEXP) {
 BEGIN_RCPP
     Rcpp::RObject __result;
     Rcpp::traits::input_parameter< int >::type nr(nrSEXP);
     __result = Rcpp::wrap(test_rnorm2(nr));
     return __result;
 END_RCPP
 }

 Generated R functions
 -------------------------------------------------------

 `.sourceCpp_13765_DLLInfo` <- dyn.load('/tmp/RtmpLKENin/sourcecpp_50641b89aeee/sourceCpp_65700.so')

 test_rnorm <- Rcpp:::sourceCppFunction(function(ncpp, nr) {}, FALSE, `.sourceCpp_13765_DLLInfo`, 'sourceCpp_13765_test_rnorm')
 test_rnorm2 <- Rcpp:::sourceCppFunction(function(nr) {}, FALSE, `.sourceCpp_13765_DLLInfo`, 'sourceCpp_13765_test_rnorm2')

 rm(`.sourceCpp_13765_DLLInfo`)

 Building shared library
 --------------------------------------------------------

 DIR: /tmp/RtmpLKENin/sourcecpp_50641b89aeee

 /usr/lib/R/bin/R CMD SHLIB -o 'sourceCpp_65700.so' --preclean  'romain.cpp'
 g++ -I/usr/share/R/include -DNDEBUG    -I"/usr/local/lib/R/site-library/Rcpp/include" -I"/tmp"    -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g  -O3 -Wall -pipe -Wno-unused -pedantic \
  -c romain.cpp -o romain.o
 g++ -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o sourceCpp_65700.so romain.o -L/usr/lib/R/lib -lR

 R>     test_rnorm(4, 2); rnorm(2)
 [[1]]
 [1]  0.539024  0.580206 -0.657503  1.554896

 [[2]]
 [1] -1.187641  0.151813

 [1] -1.08613  1.61337

 R>     set.seed(42); test_rnorm2(2); rnorm(2)
 [1]  1.370958 -0.564698
 [1] 0.363128 0.632863

 R>     set.seed(42); rnorm(2); test_rnorm2(2)
 [1]  1.370958 -0.564698
 [1] 0.363128 0.632863
 R> 
```

**Comment by romainfrancois** [comment]
This whole framework you mention is based on the assumption that only `Rcpp` calls `GetRNGState` and `PutRNGState` via the counter mechanism in the `RNGScope` class. 

When it's not the case, then you get wrong results like above. 

Not as bad as badly using `unlink`, you just get  bogus random stream. 

Also, all the `Rcpp::r*` functions know how to call `GetRNGState`, `PutRNGState` through the inheritance of the `Generator` class that itself inherits from `RNGScope`. The point of adding the `RNGScope`automagically in attributes is so that if we have a function that calls two of them, e.g.: 

```cpp
void foo(){
   NumericVector x = rnorm(3); 
   NumericVector y = runif(3) ;
}
```

we only get one call of `Get` and one call of `Put`; If attributes did not force the `RNGScope` thing in, we'd get two of each, one as the ctor/dtor pair of the class behind `rnorm` and another one related to the ctor/dtor pair of the class behind `runif`. 

So as you say above, things still work without the automatic insertion forced by attributes. You just pay for more `Get`, `Put` which is cheap anyway. 

The other benefit of the automatic handling is that it also caters for uses of the `R::rnorm`, aka `Rf_rnorm` ... calls. 

So my recommandation would be to remove the automatic `RNGScope` altogether and properly document its explicit uses: 
 - when calling `R::r*` functions
 - when calling multiple `Rcpp::r*` functions so that we pay for `Get, `Put` only once. 

I just don't believe the "if for some reason your C++ function calls Rcpp functions to generate random numbers but also an R function that generates random number, then you should use `rng=false`"

**Comment by romainfrancois** [comment]
BTW, I just don't believe that people are going to start sprinkling their code with `rng=false` to spare the cost of automatic `RNGScope`. I don't see this happening at all. But perhaps that's just personal opinion.

**Comment by jjallaire** [comment]
A few issues with removing it:

1. We'll break (or at least change the runtime behavior of) existing code

2. I don't think we can expect users to have read the documentation to the
point where they'll know to include it when they need it

3. Because PutRNGState can trigger a GC the user needs to make special
accommodation for this when returning values from a function with RNGScope
(they need to create an RObject on the stack before the RNGScope and then
assign to and return it to avoid the GC on their returned value when the
RNGScope is destructed). This is not intuitive at all and I think therefore
best done in generated code.

4. Calling R functions from within C++ code is not the design center
(especially when there are analogous C++ versions available).

I think we should keep the behavior as-is but certainly add some
documentation on this edge case.

**Comment by eddelbuettel** [comment]
It is worth considering but I'd like to hear from users who were bitten by corrupt RNG streams.  

You do point a valid issue here, I currently think it is of somewhat low likelihood in practice -- whereas what we built generally works and is widely used.   

And @jjallaire pretty much just said the same while I typed this.

**Comment by coatless** [comment]
[Additional RNG Scope Info on SO via Dirk](http://stackoverflow.com/questions/38383019/is-and-when-rngscope-scope-needed-in-rcpp-code)

```
In short:

 - _whenever_ you call the RNGs of the C API of R, you need to save and later re-set state
 - `RNGScope` automates this for you as it is so darn useful (and generally inexpensive)
 - whenever you use Rcpp Attributes, it inserts `RNGScope` (as you can see when you turn on `verbose=TRUE`

So in general you don't need to anything manual -- unless you go _really_ old school and write all code directly, foregoing the glue provided by Rcpp.

And if you use Rcpp, it doesn't matter whether you use scalar interfaces from the `R::` namespace or the vectorized `Rcpp Sugar` onces via `Rcpp::`.  `RNGScope` will be there for you.
```

**Comment by coatless** [comment]
@eddelbuettel can you close this ticket out as it repeats #823 that was resolved by #825? Or should a piece of documentation be added that emphasizes the RNG change-over and then close out this ticket?

**Comment by eddelbuettel** [comment]
C.f. #823 and #825.

**Comment by gvegayon** [comment]
Hey, I think I've found a problem related to this. In brief, an Rcpp function, call it `invoke`, that calls an R function created by Rcpp as well, call it `deterministic`, results in resetting the seed regardless on whether `deterministic` uses RNG or not. This becomes a problem when `invoke` calls a RNG function itself as it would be the case of, for example, an MCMC algorithm.

The problem goes away if, as suggested, I set `rng = false` as the following example illustrates:

```c++
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
int deterministic(int i) {
  
  return i*0;
  
}

// [[Rcpp::export(rng = false)]]
int deterministic2(int i) {
  
  return i*0;
  
}

// This function generates a random vector of standard normal variables
// [[Rcpp::export]]
NumericVector invoke(Function f, int n) {
  
  NumericVector ans(n);
  
  for (int i = 0; i < n; ++i) {
    
    // HERE IS THE KEY PART. CALLING `f` MAKES THE RANDOM SEED TO RESET
    // WHICH CAUSES THE FOLLOWING LINE TO RETURN THE SAME NUMBER IN EACH 
    // ITERATION.
    (void) f(i);
    ans[i] = norm_rand();
    
  }
  
  return ans;
  
}

/***R

deterministicR <- function(i) i*0

# Generating the random sequences
set.seed(1); ans0 <- invoke(deterministic, 10)
set.seed(1); ans1 <- invoke(deterministic2, 10)
set.seed(1); ans2 <- invoke(deterministicR, 10)

# These three columns should be the same, but they are not
# moreover, `ans0` is a constant vector.
cbind(ans0, ans1, ans2)
#             ans0       ans1       ans2
#  [1,] -0.6264538 -0.6264538 -0.6264538
#  [2,] -0.6264538  0.1836433  0.1836433
#  [3,] -0.6264538 -0.8356286 -0.8356286
#  [4,] -0.6264538  1.5952808  1.5952808
#  [5,] -0.6264538  0.3295078  0.3295078
#  [6,] -0.6264538 -0.8204684 -0.8204684
#  [7,] -0.6264538  0.4874291  0.4874291
#  [8,] -0.6264538  0.7383247  0.7383247
#  [9,] -0.6264538  0.5757814  0.5757814
# [10,] -0.6264538 -0.3053884 -0.3053884
*/
```

I was expecting all three implementations of `deterministic` to return the same result (why should the RNG seed change if `deterministic` makes no call to any RNG function?).

I was able to pin down this problem when debugging an Rcpp implementation of an MCMC algorithm. The Rcpp implementation take as argument another function (a log-likelihood function) which in my case was also written in Rcpp.

**Comment by kevinushey** [comment]
If you decide to get into the business of calling the C APIs for RNG generation, the onus is on your to explicitly call `GetRNGstate()` and `PutRNGstate()` to ensure that the C views and the R views of the RNG state are synchronized. In other words, you need:

```
GetRNGstate();
ans[i] = norm_rand();
PutRNGstate();
```

Otherwise you can (as you noted) use `rng = false`. I think this is all the expected + documented behavior.

**Comment by coatless** [comment]
@gvegayon there was a slight relapse w.r.t 12.16.0 see: 

https://github.com/RcppCore/Rcpp/issues/836

And the accompanying PR fix by @kevinushey:

https://github.com/RcppCore/Rcpp/pull/837

**Comment by kevinushey** [comment]
I didn't realize this was something we (I) had already fixed. Oops!

**Comment by coatless** [comment]
@kevinushey I hear raising a new infant does that to your brain ;-) (Congrats by the way!)

**Comment by gvegayon** [comment]
> If you want to get into the business of calling the C APIs for RNG generation directly, the onus is on you to explicitly call GetRNGstate() and PutRNGstate() to ensure that the C views and the R views of the RNG state are synchronized.

I see, indeed this is key thing to consider when using the APIs for RNG which I was not fully aware of. Thank Good for unit tests that allowed me spotting this issue! Thanks for the clarification :)!

**Comment by petesherick** [comment]
Hello, I have a Gibbs sampler where I generate a uniform and then a few sugar normals to accept/reject.  I had no idea the first normal was using the same seed as the uniform!
Example:

```
#include <RcppArmadillo.h>
using namespace Rcpp;
using namespace arma;

// [[Rcpp::export]]
List TestSampler(int number){		

	arma::vec uniforms = arma::zeros(number);
	arma::vec normals = arma::zeros(number);
	arma::vec rnorms;		
	for(int i = 0; i < number;i++){
			
		uniforms[i] = R::runif(0,1);
		
		rnorms = as<arma::vec>(rnorm(4));
		normals[i] = rnorms[0];//exactly qnorm(uniforms[i]) ??
		normals[i] = rnorms[1];
		
	}
	
	return Rcpp::List::create(Rcpp::Named("uniforms")=uniforms, Rcpp::Named("normals")=normals);

}
```

**Comment by eddelbuettel** [comment]
@petesherick Please consider deleting this post here; it seems to have little to do with the existing discussion.  Feel free to open a new issue under RcppArmadillo.

R RNG seeding is moderately complicated.  For RcppArmadillo we chose to rely on _R_ RNGs not what Armadillo does.  You can revert that.  It is (or should be ...) all documented.

**Comment by coatless** [comment]
@petesherick: It's not. There are no normal observations being generated. 

You are creating a non-initialized armadillo vector called _rnorms_. 

```cpp
arma::vec rnorms;	
```

From there, you are accessing a location in the vector that doesn't exist. 

```cpp	
rnorms = as<arma::vec>(rnorm(4));
```

In short, you should be using: `R::rnorm()` or Armadillo's [`randn()`](http://arma.sourceforge.net/docs.html#randu_randn_standalone)

**Comment by eddelbuettel** [comment]
Good correction but we are _still in the wrong thread here_.

**Comment by petesherick** [comment]
I removed all the armadillo structures I was using to store the data.  Please let me know if there is another thread with this issue or if I should open one.

**Comment by eddelbuettel** [comment]
Let's delete everything here and open a new one please.
